### ISSUE TITLE
Bug: no-unused-vars reports errors in typescript lambda types (where parameters are not variables and literally cannot be used)

### ISSUE THREAD

**Issue by pixelpax** [eslint/eslint#16974]
### Environment

Node version: 18.14
npm version: 9.0.0
Local ESLint version: 8.32.0
Global ESLint version: 8.32.0
Operating System: MacOS 13.2.1

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```
{
  "extends": ["plugin:unicorn/recommended", "next/core-web-vitals", "prettier"],

  "overrides": [
    {
      "plugins": ["testing-library", "jest"],
      "files": [
        "**/__tests__/**/*.[jt]s?(x)",
        "**/?(*.)+(spec|test).[jt]s?(x)"
      ],
      "extends": [
        "plugin:unicorn/recommended",
        "plugin:testing-library/react",
        "plugin:jest/recommended",
        "next/core-web-vitals",
        "prettier"
      ],
      "rules": {
        "jest/no-disabled-tests": "error",
        "jest/valid-expect": "error"
      }
    }
  ],
  "rules": {
    "no-unused-vars": "error",
    "testing-library/no-debugging-utils": "error",
    "react-hooks/exhaustive-deps": "error", // huge errors can creep into components don't list all of the dependencies in
    "unicorn/prefer-module": "off",
    "react/no-multi-comp": [2, { "ignoreStateless": true }],
    "unicorn/prevent-abbreviations": "off",
    "unicorn/prefer-ternary": "off", //`if` `else` syntax often leads to much more readable code
    "unicorn/no-abusive-eslint-disable": "off", // This get in the way of speed and lint-exceptions for generated files
    "unicorn/no-useless-undefined": "off", // often more descriptive to explicitly include undefined
    "unicorn/no-array-reduce": "off", // Nah, reducers are fine. Great for reducing reductions.
    "unicorn/no-negated-condition": "off",
    "unicorn/no-lonely-if": "off",
    "unicorn/filename-case": [
      "error",
      {
        "cases": {
          "camelCase": true,
          "pascalCase": true
        }
      }
    ]
  }
}
```
</details>

```ts
interface SomeThing {
  foo: (arg: string) => void // 'arg' throws unused var here, despite the fact that it is not a variable, nor is it usable 
}
```

### What did you expect to happen?

No error

### What actually happened?

There's an error

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
Hi @pixelpax, thanks for the issue!

> "no-unused-vars": "error", 

For linting TypeScript code, you should use [`@typescript-eslint/no-unused-vars`](https://typescript-eslint.io/rules/no-unused-vars) rule instead of the core `no-unused-vars` rule.
